File: robert-crypto-grpc-server/src/test/java/fr/gouv/stopc/robertserver/crypto/test/matchers/EncryptedTuplesBundleMatcher.java
package fr.gouv.stopc.robertserver.crypto.test.matchers;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.protobuf.ByteString;
import fr.gouv.stopc.robert.server.common.service.RobertClock.RobertInstant;
import fr.gouv.stopc.robert.server.common.utils.ByteUtils;
import fr.gouv.stopc.robert.server.crypto.exception.RobertServerCryptoException;
import fr.gouv.stopc.robert.server.crypto.structure.impl.CryptoAESGCM;
import fr.gouv.stopc.robertserver.crypto.test.CountryCode;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.Value;
import org.assertj.core.api.Condition;
import org.assertj.core.api.ObjectArrayAssert;

import java.io.IOException;
import java.util.Arrays;
import java.util.Base64;

import static fr.gouv.stopc.robertserver.crypto.test.KeystoreManager.*;
import static java.lang.String.format;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.fail;
import static org.assertj.core.condition.VerboseCondition.verboseCondition;

@RequiredArgsConstructor
public class EncryptedTuplesBundleMatcher {

    private final byte[] encryptedTuplesBundle;

    public static EncryptedTuplesBundleMatcher assertThatTuplesBundle(ByteString encryptedTuplesBundle) {
        return new EncryptedTuplesBundleMatcher(encryptedTuplesBundle.toByteArray());
    }

    public static Condition<EphemeralTuple> countryCode(CountryCode expectedCountryCode) {
        return verboseCondition(
                ephemeralTuple -> ephemeralTuple.getDecryptedCountryCode() == expectedCountryCode.getNumericCode(),
                "a tuple with an ECC corresponding to " + expectedCountryCode,
                ephemeralTuple -> format(
                        " but %s has an ECC corresponding to %d", ephemeralTuple,
                        ephemeralTuple.getDecryptedCountryCode()
                )
        );
    }

    public static Condition<EphemeralTuple> idA(String expectedBase64EncodedIdA) {
        return verboseCondition(
                ephemeralTuple -> expectedBase64EncodedIdA.equals(ephemeralTuple.getDecryptedBase64IdA()),
                "a tuple with an EBID corresponding to idA " + expectedBase64EncodedIdA,
                ephemeralTuple -> format(
                        " but idA of tuple %s was %s", ephemeralTuple,
                        ephemeralTuple.getDecryptedBase64IdA()
                )
        );
    }

    public static Condition<EphemeralTuple> idA(byte[] expectedIdA) {
        final var expectedBase64EncodedIdA = Base64.getEncoder().encodeToString(expectedIdA);
        return idA(expectedBase64EncodedIdA);
    }

    public static Condition<EphemeralTuple> ebidConstistentWithTupleEpoch() {
        return verboseCondition(
                ephemeralTuple -> ephemeralTuple.getEpochId() == ephemeralTuple.getDecryptedEbidEpoch(),
                "a tuple with an EBID corresponding to the unencrypted epoch",
                ephemeralTuple -> format(
                        " but %s was a tuple with an EBID for epoch %d",
                        ephemeralTuple, ephemeralTuple.getDecryptedEbidEpoch()
                )
        );
    }

    public static Condition<? super EphemeralTuple[]> aBundleWithEpochs(RobertInstant bundleStart,
            RobertInstant bundleEnd) {
        final var expectedEpochIds = bundleStart.epochsUntil(bundleEnd)
                .map(RobertInstant::asEpochId)
                .toArray(Integer[]::new);
        return verboseCondition(
                ephemeralTuples -> {
                    final var actualEpochs = Arrays.stream(ephemeralTuples)
                            .map(EphemeralTuple::getEpochId)
                            .toArray(Integer[]::new);
                    return Arrays.equals(actualEpochs, expectedEpochIds);
                },
                format("a bundle with all epochs from %s to %s", bundleStart, bundleEnd),
                ephemeralTuples -> format(" but is was a bundle containing epochs %s", Arrays.toString(ephemeralTuples))
        );
    }

    public ObjectArrayAssert<EphemeralTuple> isEncryptedWith(CryptoAESGCM cipherForTuples) {
        try {
            final var decryptedTuplesBundle = cipherForTuples.decrypt(encryptedTuplesBundle);
            final var listOfEphemeralTuples = new ObjectMapper()
                    .readValue(decryptedTuplesBundle, EphemeralTuple[].class);
            return assertThat(listOfEphemeralTuples)
                    .as("decrypted tuples bundle");
        } catch (RobertServerCryptoException e) {
            fail("tuples bundle can't be decrypted", e);
        } catch (IOException e) {
            fail("decrypted tuples bundle can't parsed", e);
        }
        return null;
    }

    @Value
    public static class EphemeralTuple {

        int epochId;

        TupleKey key;

        @Value
        public static class TupleKey {

            String ebid;

            String ecc;
        }

        @SneakyThrows
        public int getDecryptedEbidEpoch() {
            final var ebid = Base64.getDecoder().decode(key.ebid);
            final var decryptedEbid = cipherForEbidAtEpoch(epochId).decrypt(ebid);
            final var epochId = Arrays.copyOf(decryptedEbid, 3);
            return ByteUtils.convertEpoch24bitsToInt(epochId);
        }

        @SneakyThrows
        public String getDecryptedBase64IdA() {
            final var ebid = Base64.getDecoder().decode(key.ebid);
            final var decryptedEbid = cipherForEbidAtEpoch(epochId).decrypt(ebid);
            final var idA = Arrays.copyOfRange(decryptedEbid, 3, 8);
            return Base64.getEncoder().encodeToString(idA);
        }

        @SneakyThrows
        public int getDecryptedCountryCode() {
            final var ebid = Base64.getDecoder().decode(key.ebid);
            final var ecc = Base64.getDecoder().decode(key.ecc);

            // Pad to 128-bits
            final var payloadToEncrypt = Arrays.copyOf(ebid, 128 / 8);
            // AES Encryption of the payload to encrypt
            final var encryptedPayload = cipherForEcc().encrypt(payloadToEncrypt);

            // Truncate to 8 bits
            // Equivalent to MSB in ROBert spec
            final var truncatedEncryptedPayload = encryptedPayload[0];
            final var truncatedEncryptedEcc = ecc[0];

            return (truncatedEncryptedPayload ^ truncatedEncryptedEcc);
        }

        public String toString() {
            return format("EphemeralTuple(epochId=%d, key.ebid=%s, key.ecc=%s)", epochId, key.ebid, key.ecc);
        }
    }
}

File: robert-crypto-grpc-server/src/test/kotlin/fr/gouv/stopc/robertserver/crypto/test/matchers/EncryptedTuplesBundleMatcher.kt
package fr.gouv.stopc.robertserver.crypto.test.matchers

import com.google.protobuf.ByteString
import fr.gouv.stopc.robert.server.common.utils.ByteUtils
import fr.gouv.stopc.robert.server.crypto.exception.RobertServerCryptoException
import fr.gouv.stopc.robert.server.crypto.structure.impl.CryptoAESGCM
import fr.gouv.stopc.robertserver.common.RobertClock.RobertInstant
import fr.gouv.stopc.robertserver.common.base64Encode
import fr.gouv.stopc.robertserver.crypto.test.CountryCode
import fr.gouv.stopc.robertserver.crypto.test.cipherForEbidAtEpoch
import fr.gouv.stopc.robertserver.crypto.test.cipherForEcc
import fr.gouv.stopc.robertserver.crypto.test.matchers.EncryptedTuplesBundleMatcher.EphemeralTuple
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.assertj.core.api.Condition
import org.assertj.core.api.ObjectArrayAssert
import org.assertj.core.condition.VerboseCondition.verboseCondition
import java.io.IOException
import java.util.Arrays
import java.util.Base64
import kotlin.streams.toList

/**
 * Returns an [EncryptedTuplesBundleMatcher] with an operation to decrypt ephemeral tuples.
 */
fun assertThatTuplesBundle(encryptedTuplesBundle: ByteString): EncryptedTuplesBundleMatcher {
    return EncryptedTuplesBundleMatcher(encryptedTuplesBundle.toByteArray())
}

/**
 * An AssertJ [Condition] to verify the country code of an ephemeral tuple.
 */
fun countryCode(expectedCountryCode: CountryCode): Condition<EphemeralTuple> = verboseCondition(
    { ephemeralTuple -> ephemeralTuple.decryptedCountryCode == expectedCountryCode.numericCode },
    "a tuple with an ECC corresponding to $expectedCountryCode",
    { ephemeralTuple -> " but $ephemeralTuple has an ECC corresponding to ${ephemeralTuple.decryptedCountryCode}" }
)

/**
 * An AssertJ [Condition] to verify the `idA` of an ephemeral tuple.
 */
fun idA(expectedBase64EncodedIdA: String): Condition<EphemeralTuple> = verboseCondition(
    { ephemeralTuple -> expectedBase64EncodedIdA == ephemeralTuple.decryptedBase64IdA },
    "a tuple with an EBID corresponding to idA $expectedBase64EncodedIdA",
    { ephemeralTuple -> " but idA of tuple $ephemeralTuple was ${ephemeralTuple.decryptedBase64IdA}" }
)

/**
 * An AssertJ [Condition] to verify the `idA` of an ephemeral tuple.
 */
fun idA(expectedIdA: ByteArray): Condition<EphemeralTuple> = idA(expectedIdA.base64Encode())

/**
 * An AssertJ [Condition] to verify the ephemeral tuple has the same `epochId` in its cleartext epochId attribute and in the encrypted data of the `ebid`.
 */
fun ebidConstistentWithTupleEpoch(): Condition<EphemeralTuple> = verboseCondition(
    { ephemeralTuple -> ephemeralTuple.epochId == ephemeralTuple.decryptedEbidEpoch },
    "a tuple with an EBID corresponding to the unencrypted epoch",
    { ephemeralTuple -> " but $ephemeralTuple was a tuple with an EBID for epoch ${ephemeralTuple.decryptedEbidEpoch}" }
)

/**
 * An AssertJ [Condition] to verify a list contains `ebid`s for a given continuous period.
 */
fun aBundleWithEpochs(bundleStart: RobertInstant, bundleEnd: RobertInstant): Condition<Array<EphemeralTuple>> {
    val expectedEpochIds = bundleStart.epochsUntil(bundleEnd)
        .map(RobertInstant::asEpochId)
        .toList()
    return verboseCondition(
        { ephemeralTuples ->
            val actualEpochs = ephemeralTuples
                .map { it.epochId }
            actualEpochs == expectedEpochIds
        },
        "a bundle with all epochs from $bundleStart to $bundleEnd",
        { ephemeralTuples -> " but is was a bundle containing epochs:[\n${ephemeralTuples.joinToString(",\n  ")}]" }
    )
}

class EncryptedTuplesBundleMatcher(private val encryptedTuplesBundle: ByteArray) {

    /**
     * Decrypts the encrypted bundle of ephemeral tuples.
     * @return a ListAssert verifiable using conditions [countryCode], [idA], [ebidConstistentWithTupleEpoch], ...
     */
    fun isEncryptedWith(cipherForTuples: CryptoAESGCM): ObjectArrayAssert<EphemeralTuple> {
        try {
            val decryptedTuplesBundle = cipherForTuples.decrypt(encryptedTuplesBundle)
                .decodeToString()
            val listOfEphemeralTuples = Json.decodeFromString<Array<EphemeralTuple>>(decryptedTuplesBundle)
            return assertThat(listOfEphemeralTuples)
                .describedAs("decrypted tuples bundle")
        } catch (e: RobertServerCryptoException) {
            fail<Any>("tuples bundle can't be decrypted", e)
        } catch (e: IOException) {
            fail<Any>("decrypted tuples bundle can't parsed", e)
        }
        throw IllegalStateException()
    }

    @Serializable
    data class EphemeralTuple(
        val epochId: Int,
        val key: TupleKey
    ) {

        @Serializable
        data class TupleKey(
            val ebid: String,
            val ecc: String
        )

        val decryptedEbidEpoch: Int
            get() {
                val ebid = Base64.getDecoder().decode(key.ebid)
                val decryptedEbid = cipherForEbidAtEpoch(epochId).decrypt(ebid)
                val epochId = Arrays.copyOf(decryptedEbid, 3)
                return ByteUtils.convertEpoch24bitsToInt(epochId)
            }

        val decryptedBase64IdA: String
            get() {
                val ebid = Base64.getDecoder().decode(key.ebid)
                val decryptedEbid = cipherForEbidAtEpoch(epochId).decrypt(ebid)
                val idA = Arrays.copyOfRange(decryptedEbid, 3, 8)
                return Base64.getEncoder().encodeToString(idA)
            }

        val decryptedCountryCode: Int
            get() {
                val ebid = Base64.getDecoder().decode(key.ebid)
                val ecc = Base64.getDecoder().decode(key.ecc)

                // Pad to 128-bits
                val payloadToEncrypt = Arrays.copyOf(ebid, 128 / 8)
                // AES Encryption of the payload to encrypt
                val encryptedPayload = cipherForEcc().encrypt(payloadToEncrypt)

                // Truncate to 8 bits
                // Equivalent to MSB in ROBert spec
                val truncatedEncryptedPayload = encryptedPayload[0]
                val truncatedEncryptedEcc = ecc[0]
                return truncatedEncryptedPayload.toInt() xor truncatedEncryptedEcc.toInt()
            }
    }
}
